tket2/src/rewrite_space/factory/kak.rs

```diff
@@ -0,0 +1,255 @@
+                        // anything as this is not a 2q sub circuit.
+                        if two_qubit_node == expanded_node {
+
+                            searching_walker.pattern.wires.push(expanded_wire);
+
+                            matches.push((searching_walker.pattern, searching_walker.walker));
+
+                            searching_walker.two_qubit_node = None;
+                            searching_walker.frontier_wires = vec![
+                                expanded_walker.get_wire(two_qubit_node, OutgoingPort::from(0)),
+                                expanded_walker.get_wire(two_qubit_node, OutgoingPort::from(1)),
+                            ];
+                            searching_walker.walker = expanded_walker;
+
+                            searching_walker_list.push(searching_walker);
+                        }
+
+                    // If we have not found this 2q gate before then we need
+                    // to explore the other wire to see if it meets at
+                    // the same place. As such we put the searcher
+                    // back on the queue.
+                    } else {
+                        searching_walker.two_qubit_node = Some(expanded_node);
+                        searching_walker_list.push(searching_walker);
+                    }
+                }
+            }
+
+        }
+
+        matches.into_iter()
+    }
+
+    fn op_cost(&self, op: &hugr::ops::OpType) -> Option<Self::Cost> {
+        todo!()
+    }
+    
+    fn get_replacement(
+        &self,
+        pattern_match: &Self::PatternMatch,
+        host: &PersistentHugr,
+    ) -> Option<Hugr> {
+        todo!()
+    }
+    
+    fn map_boundary(
+        &self,
+        node: PatchNode,
+        port: hugr::Port,
+        pattern_match: &Self::PatternMatch,
+        host: &PersistentHugr,
+    ) -> hugr::Port {
+        todo!()
+    }
+    
+    fn get_name(&self, pattern_match: &Self::PatternMatch, host: &PersistentHugr) -> String {
+        todo!()
+    }
+}
+
+/// only call this on nodes that are known quantum gates
+fn n_qubits<N: Copy>(node: N, hugr: &impl HugrView<Node = N>) -> usize {
+    debug_assert_eq!(
+        hugr.signature(node).unwrap().input_count(),
+        hugr.signature(node).unwrap().output_count(),
+        "expected a quantum gate"
+    );
+    hugr.signature(node).unwrap().input_count()
+}
+
+fn empty_2qb_hugr() -> Hugr {
+    let builder = DFGBuilder::new(endo_sig(vec![qb_t(), qb_t()])).unwrap();
+    let [q0, q1] = builder.input_wires_arr();
+    builder.finish_hugr_with_outputs(vec![q0, q1]).unwrap()
+}
+
+fn one_cz_2qb_hugr() -> Hugr {
+    let mut builder = DFGBuilder::new(endo_sig(vec![qb_t(), qb_t()])).unwrap();
+    let [q0, q1] = builder.input_wires_arr();
+    let cz1 = builder.add_dataflow_op(Tk2Op::CZ, vec![q0, q1]).unwrap();
+    let [q0, q1] = cz1.outputs_arr();
+    builder.finish_hugr_with_outputs(vec![q0, q1]).unwrap()
+}
+
+#[test]
+fn test_kak_resynthesis_init() {
```

Comment:
not important but useful to know:

if you replace `#[test]` with `#[rstest]` (you need to import it with `use rstest::rstest`) and you add `#[fixture]` (also imported from `rstest`) in front of your helper function `one_cz_2qb_hugr`, then you can write the test as

```rust
#[rstest]
fn test_kak_resynthesis_init(one_cz_2qb_hugr: Hugr) { ... }
```

It's just syntactic sugar to move the construction of the `Hugr` from a function call in your function body as you have it now, to something that looks like an argument passed to the test.